File: languages/kotlin/FatigueCalculator.kt
// FatigueCalculator.kt
// Kotlin implementation of fatigue calculation algorithms
// Supports JVM, Android, and Kotlin/Native targets

package fatigue

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.sqrt

@Serializable
enum class ActivityLevel {
    SEDENTARY,
    ACTIVE,
    HIGH_INTENSITY
}

@Serializable
enum class FatigueLevel {
    MILD,
    MODERATE,
    SEVERE
}

class FatigueException(message: String) : Exception(message)

@Serializable
data class FatigueProfile(
    val sedentaryLimit: Double = 5.0,
    val activeLimit: Double = 8.0,
    val highIntensityLimit: Double = 10.0
) {
    fun getLimit(level: ActivityLevel): Double = when (level) {
        ActivityLevel.SEDENTARY -> sedentaryLimit
        ActivityLevel.ACTIVE -> activeLimit
        ActivityLevel.HIGH_INTENSITY -> highIntensityLimit
    }
}

@Serializable
data class FatigueSummary(
    val total: Double,
    val count: Int,
    val average: Double,
    val maximum: Double,
    val minimum: Double,
    val recovered: Boolean
) {
    companion object {
        fun empty(): FatigueSummary = FatigueSummary(
            total = 0.0,
            count = 0,
            average = 0.0,
            maximum = 0.0,
            minimum = 0.0,
            recovered = false
        )
    }
}

class FatigueCalculator(private val profile: FatigueProfile = FatigueProfile()) {
    
    init {
        if (profile.sedentaryLimit <= 0 || profile.activeLimit <= 0 || profile.highIntensityLimit <= 0) {
            throw FatigueException("Limits must be positive")
        }
    }
    
    fun calculateFatigue(score: Double, level: ActivityLevel): Result<Double> {
        if (score < 0) {
            return Result.failure(FatigueException("Score cannot be negative"))
        }
        
        val limit = profile.getLimit(level)
        return Result.success(minOf(score, limit))
    }
    
    fun calculateRecovery(fatigue: Double, sleep: Double): Result<Double> {
        if (fatigue < 0 || sleep < 0) {
            return Result.failure(FatigueException("Values cannot be negative"))
        }
        
        return Result.success(maxOf(fatigue - (sleep * 0.1), 0.0))
    }
    
    fun summarize(scores: List<Double>): Result<FatigueSummary> {
        if (scores.isEmpty()) {
            return Result.failure(FatigueException("Cannot summarize empty list"))
        }
        
        val total = scores.sum()
        val count = scores.size
        val average = total / count
        val maximum = scores.maxOrNull() ?: 0.0
        val minimum = scores.minOrNull() ?: 0.0
        val recovered = average < 5.0
        
        return Result.success(
            FatigueSummary(
                total = total,
                count = count,
                average = average,
                maximum = maximum,
                minimum = minimum,
                recovered = recovered
            )
        )
    }
    
    fun computeIntensityMetrics(base: Double, multiplier: Double): Result<Double> {
        if (multiplier < 0) {
            return Result.failure(FatigueException("Multiplier cannot be negative"))
        }
        return Result.success(base * multiplier)
    }
    
    fun adjustForActivityLevel(base: Double, activity: ActivityLevel): Result<Double> =
        calculateFatigue(base, activity)
    
    fun verifyThreshold(computed: Double, threshold: Double): Result<Boolean> {
        if (computed < 0 || threshold < 0) {
            return Result.failure(FatigueException("Values cannot be negative"))
        }
        return Result.success(computed >= threshold)
    }
    
    fun metricsToJSON(summary: FatigueSummary): String {
        return Json.encodeToString(FatigueSummary.serializer(), summary)
    }
    
    fun metricsFromJSON(json: String): Result<FatigueSummary> {
        return try {
            Result.success(Json.decodeFromString(FatigueSummary.serializer(), json))
        } catch (e: Exception) {
            Result.failure(FatigueException("Failed to parse JSON: ${e.message}"))
        }
    }
    
    fun serializeMetrics(summary: FatigueSummary): String = metricsToJSON(summary)
    
    fun categorizeFatigueLevel(score: Double): Result<FatigueLevel> {
        return when {
            score < 0 -> Result.failure(FatigueException("Score cannot be negative"))
            score < 3.0 -> Result.success(FatigueLevel.MILD)
            score < 7.0 -> Result.success(FatigueLevel.MODERATE)
            else -> Result.success(FatigueLevel.SEVERE)
        }
    }
    
    fun computeZScore(values: List<Double>): Result<Double> {
        if (values.isEmpty()) {
            return Result.failure(FatigueException("Cannot compute z-score for empty list"))
        }
        
        val mean = values.average()
        val variance = values.map { (it - mean) * (it - mean) }.average()
        val stdDev = sqrt(variance)
        
        return if (stdDev == 0.0) {
            Result.success(0.0)
        } else {
            Result.success((values.first() - mean) / stdDev)
        }
    }
    
    fun calculateRMSE(predicted: List<Double>, actual: List<Double>): Result<Double> {
        if (predicted.size != actual.size || predicted.isEmpty()) {
            return Result.failure(FatigueException("Lists must have same non-zero length"))
        }
        
        val mse = predicted.zip(actual) { p, a -> (p - a) * (p - a) }.average()
        return Result.success(sqrt(mse))
    }
    
    fun generateForecastTrend(historical: List<Double>, periods: Int): Result<List<Double>> {
        if (historical.isEmpty() || periods <= 0) {
            return Result.failure(FatigueException("Historical data required and periods must be positive"))
        }
        
        val last = historical.last()
        val trend = (historical.lastOrNull() ?: last) - (historical.firstOrNull() ?: last)
        
        val forecasts = (1..periods).map { i ->
            maxOf(0.0, last + (trend * i))
        }
        
        return Result.success(forecasts)
    }
    
    // Coroutine-based async versions for Kotlin concurrency
    suspend fun calculateFatigueAsync(score: Double, level: ActivityLevel): Result<Double> {
        return calculateFatigue(score, level)
    }
    
    suspend fun summarizeAsync(scores: List<Double>): Result<FatigueSummary> {
        return summarize(scores)
    }
}

File: languages/swift/FatigueCalculator.swift
// FatigueCalculator.swift
// Swift implementation of fatigue calculation algorithms
// Ported from Rust reference implementation

import Foundation

public struct FatigueProfile: Hashable, Equatable {
    let sedentaryLimit: Decimal
    let activeLimit: Decimal
    let highIntensityLimit: Decimal
    
    func getLimit(for activity: ActivityLevel) -> Decimal {
        switch activity {
        case .sedentary: return sedentaryLimit
        case .active: return activeLimit
        case .highIntensity: return highIntensityLimit
        }
    }
}

public enum ActivityLevel: String, CaseIterable {
    case sedentary = "sedentary"
    case active = "active"
    case highIntensity = "high_intensity"
}

public enum FatigueError: Error {
    case negativeScore
    case invalidLevel
    case emptySlice
}

public struct FatigueSummary {
    let total: Decimal
    let count: Int
    let average: Decimal
    let maximum: Decimal
    let minimum: Decimal
    let recovered: Bool
}

public class FatigueCalculator {
    public let profile: FatigueProfile
    
    public init(profile: FatigueProfile) {
        self.profile = profile
    }
    
    public func calculateFatigue(score: Decimal, level: ActivityLevel) -> Result<Decimal, FatigueError> {
        if score < 0 {
            return .failure(.negativeScore)
        }
        
        let limit = profile.getLimit(for: level)
        return .success(min(score, limit))
    }
    
    public func calculateFatigue(score: Decimal, level: ActivityLevel) throws -> Decimal {
        if score < 0 {
            throw FatigueError.negativeScore
        }
        
        let limit = profile.getLimit(for: level)
        return min(score, limit)
    }
    
    public func calculateRecovery(fatigue: Decimal, sleep: Decimal) -> Result<Decimal, FatigueError> {
        if fatigue < 0 {
            return .failure(.negativeScore)
        }
        if sleep < 0 {
            return .failure(.negativeScore)
        }
        
        return .success(max(fatigue - (sleep * 0.1), 0))
    }
    
    public func summarize(scores: [Decimal]) -> Result<FatigueSummary, FatigueError> {
        if scores.isEmpty {
            return .failure(.emptySlice)
        }
        
        let total = scores.reduce(0) { $0 + $1 }
        let count = scores.count
        let average = total / Decimal(count)
        let maximum = scores.max() ?? 0
        let minimum = scores.min() ?? 0
        let recovered = average < 5
        
        return .success(FatigueSummary(
            total: total,
            count: count,
            average: average,
            maximum: maximum,
            minimum: minimum,
            recovered: recovered
        ))
    }
    
    public func computeIntensityMetrics(base: Decimal, multiplier: Decimal) -> Result<Decimal, FatigueError> {
        if multiplier < 0 {
            return .failure(.invalidLevel)
        }
        
        return .success(base * multiplier)
    }
    
    public func adjustForActivityLevel(base: Decimal, activity: ActivityLevel) -> Result<Decimal, FatigueError> {
        let adjustmentFactors: [ActivityLevel: Decimal] = [
            .sedentary: Decimal(1.0),
            .active: Decimal(1.2),
            .highIntensity: Decimal(1.5)
        ]
        
        guard let factor = adjustmentFactors[activity] else {
            return .failure(.invalidLevel)
        }
        
        return .success(base * factor)
    }
    
    public func verifyThreshold(computed: Decimal, threshold: Decimal) -> Result<Bool, FatigueError> {
        if computed < 0 || threshold < 0 {
            return .failure(.negativeScore)
        }
        return .success(computed >= threshold)
    }
    
    public func metricsToJSON(metrics: FatigueSummary) -> Data? {
        let dict: [String: Any] = [
            "total": metrics.total as NSDecimalNumber,
            "count": metrics.count,
            "average": metrics.average as NSDecimalNumber,
            "max": metrics.maximum as NSDecimalNumber,
            "min": metrics.minimum as NSDecimalNumber,
            "recovered": metrics.recovered
        ]
        return try? JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted)
    }
    
    public func metricsFromJSON(data: Data) -> FatigueSummary? {
        guard let dict = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
            return nil
        }
        
        guard let total = dict["total"] as? NSNumber,
              let count = dict["count"] as? Int,
              let average = dict["average"] as? NSNumber,
              let max = dict["max"] as? NSNumber,
              let min = dict["min"] as? NSNumber,
              let recovered = dict["recovered"] as? Bool else {
            return nil
        }
        
        return FatigueSummary(
            total: Decimal(total.doubleValue),
            count: count,
            average: Decimal(average.doubleValue),
            maximum: Decimal(max.doubleValue),
            minimum: Decimal(min.doubleValue),
            recovered: recovered
        )
    }
    
    public func serializeMetrics(metrics: FatigueSummary) -> String? {
        guard let data = metricsToJSON(metrics: metrics),
              let json = String(data: data, encoding: .utf8) else {
            return nil
        }
        return json
    }
    
    public func categorizeFatigueLevel(score: Decimal) -> Result<String, FatigueError> {
        switch score {
        case ..<3: return .success("mild")
        case 3..<7: return .success("moderate")
        case 7..<Integer.max: return .success("severe")
        default: return .failure(.negativeScore)
        }
    }
    
    public func computeZScore(values: [Decimal]) -> Result<Decimal, FatigueError> {
        if values.isEmpty {
            return .failure(.emptySlice)
        }
        
        let mean = values.reduce(0, +) / Decimal(values.count)
        let variance = values.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Decimal(values.count)
        let stdDev = Decimal(Double(variance.doubleValue).squareRoot())
        
        if stdDev == 0 {
            return .success(0)
        }
        
        return .success((mean - mean) / stdDev)
    }
    
    public func calculateRMSE(predicted: [Decimal], actual: [Decimal]) -> Result<Decimal, FatigueError> {
        if predicted.count != actual.count || predicted.isEmpty {
            return .failure(.emptySlice)
        }
        
        let squaredErrors = zip(predicted, actual).map { ($0 - $1) * ($0 - $1) }
        let mse = squaredErrors.reduce(0, +) / Decimal(predicted.count)
        return .success(Decimal(Double(mse.doubleValue).squareRoot()))
    }
    
    public func generateForecastTrend(historical: [Decimal], periods: Int) -> Result<[Decimal], FatigueError> {
        if historical.isEmpty || periods <= 0 {
            return .failure(.emptySlice)
        }
        
        guard let lastValue = historical.last else {
            return .failure(.emptySlice)
        }
        
        var forecasts: [Decimal] = []
        for i in 1...periods {
            let forecast = lastValue + Decimal(i)
            forecasts.append(forecast)
        }
        
        return .success(forecasts)
    }
}
